File: 2024/b.py
#!/usr/bin/env python3

# ICPC Domestic Qualifier
# Problem B - Overtaking

def accum(a):
    ac = [a[0]]
    for i in range(1, len(a)):
        ac.append(ac[-1] + a[i])
    return ac

def sign(ab):
    ai, bi = ab
    return -1 if ai < bi else 1

def solve(a, b):
    acc_a = accum(a)
    acc_b = accum(b)
    pairs = [(ai, bi) for ai, bi in zip(acc_a, acc_b) if ai != bi]
    ovtks = []
    for ai, bi in pairs:
        if ovtks == [] or sign(ovtks[-1]) != sign((ai, bi)):
            ovtks.append((ai, bi))
    return len(ovtks) - 1

while True:
    n = int(input())
    if n == 0:
        break
    a = list(map(int, input().split()))
    b = list(map(int, input().split()))
    assert len(a) == len(b) == n
    print(solve(a, b))

File: 2024/b.hs
#!/usr/bin/env runghc

-- ICPC Domestic Qualifier
-- Problem B - Overtaking

-- 入力を受け取って結果をprintするだけのmain関数
main :: IO ()
main = do n <- read <$> getLine
          if n == 0
             then return ()
             else do as <- map read . words <$> getLine
                     bs <- map read . words <$> getLine
                     print $ solve as bs
                     main

-- アルゴリズム本体
solve :: [Int] -> [Int] -> Int
solve as bs =
    -- 符号が変化する位置の個数
    length [1 | (s1, s2) <- zip signs (tail signs), s1 /= s2]
  where
    acc_a = scanl (+) 0 as  -- 累積和のリスト
    acc_b = scanl (+) 0 bs  -- 〃
    -- 差の符号のリスト（ただし0を除去）
    signs = [signum (a - b) | (a, b) <- zip acc_a acc_b, a /= b]
